arangod/Pregel/Algos/AIR/Greenspun/Primitives.cpp

```diff
@@ -0,0 +1,1264 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2020 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Heiko Kernbach
+/// @author Lars Maier
+/// @author Markus Pfeiffer
+////////////////////////////////////////////////////////////////////////////////
+
+#include <Basics/VelocyPackHelper.h>
+#include <velocypack/Collection.h>
+#include <velocypack/Iterator.h>
+#include <velocypack/velocypack-aliases.h>
+
+#include <iostream>
+#include <list>
+
+#include "Extractor.h"
+#include "Interpreter.h"
+#include "Primitives.h"
+
+namespace arangodb::greenspun {
+
+EvalResult Prim_Min(Machine& ctx, VPackSlice const params, VPackBuilder& result) {
+  bool set = false;
+  auto tmp = double{0};
+  for (auto p : VPackArrayIterator(params)) {
+    if (p.isNumber<double>()) {
```

Comment:
Is it intentional that this only works on double numbers? I.e. this does not support any integers in the VPack input array.

Reply:
It does, because `isNumber<double>` ignores the first branch `std::is_integral<T>` (`T = double`) and then returns the result of `isNumber` which is `isInteger() || isDouble()`.

Also I'm pretty sure that I have evaluated `["min", 1, 2]` and it worked fine.

Reply:
Seems like it's all doubles:
```
air> ["max", 9007199254740992, 9007199254740993]
 = 9007199254740992
air> ["max", 9007199254740992, 9007199254740994]
 = 9007199254740994
```
(double mantissa is 53 bits, and 2^53 = 9,007,199,254,740,992)

Reply:
Wait, is this discussion about the fact that we represent numbers internally as doubles during evaluation or about whether the code is correct, i.e. it does/does not break when a number is not represented as double in vpack?

Reply:
I think this discussion is resolved, please _unresolve_ if it's not.